fileAttributeCaching/src/main/kotlin/com/pkware/java/nio/file/fileAttributeCaching/FileAttributeCachingFileSystemProvider.kt

```diff
@@ -0,0 +1,348 @@
+        fileSystems.computeIfAbsent(uri) {
+            FileAttributeCachingFileSystem(
+                env.getValue("filesystem") as FileSystem,
+                this
+            )
+        }
+
+    override fun getFileSystem(uri: URI): FileSystem =
+        fileSystems[uri] ?: throw FileSystemNotFoundException("Filesystem for $uri not found")
+
+    override fun getPath(uri: URI): Path = Path.of(uri)
+
+    override fun newByteChannel(
+        path: Path,
+        options: MutableSet<out OpenOption>,
+        vararg attrs: FileAttribute<*>?
+    ): SeekableByteChannel = Files.newByteChannel(path.asCachingPath().delegate, options, *attrs)
+
+    override fun newDirectoryStream(
+        dir: Path,
+        filter: DirectoryStream.Filter<in Path>
+    ): DirectoryStream<Path> = Files.newDirectoryStream(dir.asCachingPath().delegate, filter)
+
+    override fun createDirectory(
+        dir: Path,
+        vararg attrs: FileAttribute<*>
+    ) {
+        Files.createDirectory(dir.asCachingPath().delegate, *attrs)
+    }
+
+    override fun delete(path: Path) {
+        Files.delete(path.asCachingPath().delegate)
+    }
+
+    /**
+     * Copies [source] to [target] with the specified [options].
+     *
+     * If the [options] include [StandardCopyOption.COPY_ATTRIBUTES] and both the [source] and [target] are
+     * [FileAttributeCachingPath]s, then the cached attributes are copied from the [source] to the [target].
+     *
+     * See [FileSystemProvider.copy] for details.
+     *
+     * @param source The [Path] to copy from.
+     * @param target The destination [Path] to copy to.
+     * @param options The [CopyOption]s to use during this copy operation, can be `null`.
+     */
+    @Suppress("SpreadOperator")
+    override fun copy(source: Path, target: Path, vararg options: CopyOption?) {
+        val delegateSourcePath = source.asCachingPath().delegate
+        val delegateTargetPath = target.asCachingPath().delegate
+
+        if (options.contains(StandardCopyOption.COPY_ATTRIBUTES)) {
+            // filter out StandardCopyOption.COPY_ATTRIBUTES here
+            val newOptions = options.filter {
+                it != StandardCopyOption.COPY_ATTRIBUTES
+            }.toTypedArray()
+
+            // if we have both target and source caching paths we copy the attributes from source to target and run
+            // Files.copy(source, target, *newOptions) during copyCachedAttributesTo
+            if (source is FileAttributeCachingPath && target is FileAttributeCachingPath) {
+                source.copyCachedAttributesTo(target) {
+                    Files.copy(delegateSourcePath, delegateTargetPath, *newOptions)
+                }
+            } else {
+                Files.copy(delegateSourcePath, delegateTargetPath, *options)
+            }
+        } else {
+            // if we are not StandardCopyOption.COPY_ATTRIBUTES we dont care about caching the attributes for the
+            // copied file.
```

Comment:
This whole comment doesn't read well for me. Maybe: 
```suggestion
            // If the StandardCopyOption.COPY_ATTRIBUTES option is not selected, there is no need to cache the 
            // attributes for the copied file.
```

Reply:
This exact comment shows up a couple times.